core/module_test.cc

```diff
@@ -0,0 +1,201 @@
+                                ...) {
+  return nullptr;
+}
+
+void *mem_alloc(size_t size) {
+  void *ptr = malloc(size);
+
+  if (ptr) memset(ptr, 0, size);
+
+  return ptr;
+}
+
+void *mem_realloc(void *ptr, size_t size) { return realloc(ptr, size); }
+
+void mem_free(void *ptr) { free(ptr); }
+
+// -------------------------------------------------------------------------
+
+class AcmeModule : public Module {
+  public:
+    AcmeModule() : Module() {}
+
+    static const gate_idx_t kNumIGates = 1;
+    static const gate_idx_t kNumOGates = 1;
+
+    static const Commands<Module> cmds;
+
+    void Foo(struct snobj *arg) { n += 1; }
+
+    int n = {};
+};
+
+const Commands<Module> AcmeModule::cmds = {
+  {"foo", MODULE_FUNC &AcmeModule::Foo, 0}
+};
+
+int create_acme(const char *name, Module **m) {
+  EXPECT_EQ(1, ModuleBuilder::all_module_builders().size());
+  EXPECT_EQ(1, ModuleBuilder::all_module_builders().count("AcmeModule"));
+
+  const ModuleBuilder &builder = ModuleBuilder::all_module_builders().find("AcmeModule")->second;
+
+  std::string mod_name;
+  if (name) {
+    if (ModuleBuilder::all_modules().count(name)) {
+      return EEXIST;
+    }
+    mod_name = name;
+  } else {
+    mod_name = ModuleBuilder::GenerateDefaultName(builder.class_name(), builder.name_template());
+  }
+
+  *m = builder.CreateModule(mod_name);
+  builder.AddModule(*m); 
+
+  EXPECT_EQ("AcmeModule", builder.class_name());
+  EXPECT_EQ("acme_module", builder.name_template());
+  EXPECT_EQ("foo bar", builder.help_text());
+  EXPECT_EQ(1, builder.cmds().size());
+
+  return 0;
+}
+
+// Check that new module classes are actually created correctly and stored in
+// the table of module classes
+TEST(ModuleBuilderTest, RegisterModuleClass) {
+  ADD_MODULE(AcmeModule, "acme_module", "foo bar")
+  ASSERT_TRUE(__module__AcmeModule);
+
+  EXPECT_EQ(1, ModuleBuilder::all_module_builders().size());
+  EXPECT_EQ(1, ModuleBuilder::all_module_builders().count("AcmeModule"));
+
+  const ModuleBuilder &builder = ModuleBuilder::all_module_builders().find("AcmeModule")->second;
+  EXPECT_EQ("AcmeModule", builder.class_name());
+  EXPECT_EQ("acme_module", builder.name_template());
+  EXPECT_EQ("foo bar", builder.help_text());
+  EXPECT_EQ(1, builder.NumIGates());
+  EXPECT_EQ(1, builder.NumOGates());
+  EXPECT_EQ(1, builder.cmds().size());
+
+  ModuleBuilder::all_module_builders_holder(true);
+}
+
+// Check that module builders create modules correctly when given a name
+TEST(ModuleBuilderTest, CreateModuleWithName) {
+  Module *m1, *m2;
+  ADD_MODULE(AcmeModule, "acme_module", "foo bar")
+  ASSERT_TRUE(__module__AcmeModule);
+
+  EXPECT_EQ(0, create_acme("bar", &m1));
+  ASSERT_NE(nullptr, m1);
+  EXPECT_EQ(1, ModuleBuilder::all_modules().size());
+  EXPECT_EQ(EEXIST, create_acme("bar", &m2));
+  EXPECT_EQ(1, ModuleBuilder::all_modules().count("bar"));
+
+  ModuleBuilder::DestroyAllModules();
+  ModuleBuilder::all_module_builders_holder(true);
```

Comment:
You might consider creating a test fixture and using TEST_F; then you can put these common teardown routines in the fixture's TearDown() method.  (And any common setup routines can go in SetUp().)